k2/csrc/connect.cu

```diff
@@ -0,0 +1,503 @@
+          int32_t fsa_idx0 = ans_row_ids1_data[i],
+                  start_state = fsas_row_splits1_data[fsa_idx0];
+          // If the following fails, it likely means an input FSA was invalid
+          // (e.g. had exactly one state, which is not allowed).  Either that,
+          // or a code error.
+          K2_DCHECK_LT(start_state, fsas_row_splits1_data[fsa_idx0 + 1]);
+          ans_data[i] = start_state;
+        });
+    return ans;
+  }
+
+
+  /*
+    Returns the final batch of states.  This will include all final-states that
+    existed in the original FSAs.
+   */
+  std::unique_ptr<Ragged<int32_t>> GetFinalBatch() {
+    NVTX_RANGE(K2_FUNC);
+    int32_t num_fsas = NumFsas();
+    const int32_t *fsas_row_splits1_data = fsas_.RowSplits(1).Data();
+    Array1<int32_t> has_final_state(c_, num_fsas + 1);
+    int32_t *has_final_state_data = has_final_state.Data();
+    K2_EVAL(
+        c_, num_fsas, lambda_set_has_final_state, (int32_t i)->void {
+          int32_t split = fsas_row_splits1_data[i],
+                  next_split = fsas_row_splits1_data[i + 1];
+          has_final_state_data[i] = (next_split > split);
+        });
+    ExclusiveSum(has_final_state, &has_final_state);
+
+    int32_t n = has_final_state[num_fsas];
+    std::unique_ptr<Ragged<int32_t>> ans = std::make_unique<Ragged<int32_t>>(
+        RaggedShape2(&has_final_state, nullptr, n), Array1<int32_t>(c_, n));
+    int32_t *ans_data = ans->values.Data();
+    const int32_t *ans_row_ids1_data = ans->RowIds(1).Data();
+    K2_EVAL(
+        c_, n, lambda_set_final_state, (int32_t i)->void {
+          int32_t fsa_idx0 = ans_row_ids1_data[i],
+                  final_state = fsas_row_splits1_data[fsa_idx0 + 1] - 1;
+          // If the following fails, it likely means an input FSA was invalid
+          // (e.g. had exactly one state, which is not allowed).  Either that,
+          // or a code error.
+          K2_DCHECK_GT(final_state, fsas_row_splits1_data[fsa_idx0]);
+          ans_data[i] = final_state;
+        });
+    return ans;
+  }
+
+  /* Does the main work of connecting and returns the resulting FSAs.
+        @param [out] arc_map  if non-NULL, the map from (arcs in output)
+                     to (corresponding arcs in input) is written to here.
+        @return   Returns the connected FsaVec.
+   */
+  FsaVec Connect(Array1<int32_t> *arc_map) {
+    NVTX_RANGE(K2_FUNC);
+    Array1<int32_t> dest_states_idx01 = GetDestStates(fsas_, true);
+    dest_states_ = Ragged<int32_t>(fsas_.shape, dest_states_idx01);
+    incoming_arcs_ = GetIncomingArcs(fsas_, dest_states_idx01);
+
+    // Mark accessible states
+    std::unique_ptr<Ragged<int32_t>> iter = GetStartBatch();
+    while (iter != nullptr)
+      iter = GetNextBatch(*iter);
+
+    // Mard coaccessible states
+    std::unique_ptr<Ragged<int32_t>> riter = GetFinalBatch();
+    while (riter != nullptr)
+      riter = GetNextBatchBackward(*riter);
```

Comment:
The forward and backward passes are independent.
Can we use a thread pool to process them concurrently?
Please refer to
https://github.com/k2-fsa/k2/blob/926f160f3d4dc986b5f556852f65cbaebb78a717/k2/csrc/intersect_dense_pruned.cu#L270
for an example.